src/ethereum/prague/requests.py

```diff
@@ -0,0 +1,207 @@
+"""
+Requests were introduced in EIP-7685 as a a general purpose framework for
+storing contract-triggered requests. It extends the execution header and
+body with a single field each to store the request information.
+This inherently exposes the requests to the consensus layer, which can
+then process each one.
+
+[EIP-7685]: https://eips.ethereum.org/EIPS/eip-7685
+"""
+
+from dataclasses import dataclass
+from typing import Tuple, Union
+
+from .. import rlp
+from ..base_types import (
+    U64,
+    Bytes,
+    Bytes32,
+    Bytes48,
+    Bytes96,
+    slotted_freezable,
+)
+from .blocks import Receipt, decode_receipt
+from .fork_types import Address
+from .utils.hexadecimal import hex_to_address
+
+DEPOSIT_CONTRACT_ADDRESS = hex_to_address(
+    "0x00000000219ab540356cbb839cbe05303d7705fa"
+)
+DEPOSIT_REQUEST_TYPE = b"\x00"
+WITHDRAWAL_REQUEST_TYPE = b"\x01"
+CONSOLIDATION_REQUEST_TYPE = b"\x02"
+WITHDRAWAL_REQUEST_LENGTH = 76
+CONSOLIDATION_REQUEST_LENGTH = 116
+
+
+@slotted_freezable
+@dataclass
+class DepositRequest:
+    """
+    Requests for validator deposits on chain (See [EIP-6110]).
+
+    [EIP-6110]: https://eips.ethereum.org/EIPS/eip-6110
+    """
+
+    public_key: Bytes48
+    withdrawal_credentials: Bytes32
+    amount: U64
+    signature: Bytes96
+    index: U64
+
+
+@slotted_freezable
+@dataclass
+class WithdrawalRequest:
+    """
+    Requests for execution layer withdrawals (See [EIP-7002]).
+
+    [EIP-7002]: https://eips.ethereum.org/EIPS/eip-7002
+    """
+
+    source_address: Address
+    validator_public_key: Bytes48
+    amount: U64
+
+
+@slotted_freezable
+@dataclass
+class ConsolidationRequest:
+    """
+    Requests for validator consolidation (See [EIP-7251]).
+
+    [EIP-7251]: https://eips.ethereum.org/EIPS/eip-7251
+    """
+
+    source_address: Address
+    source_public_key: Bytes48
+    target_public_key: Bytes48
+
+
+Request = Union[DepositRequest, WithdrawalRequest, ConsolidationRequest]
+
+
+def encode_request(req: Request) -> Bytes:
+    """
+    Serialize a `Request` into a byte sequence.
+
+    `Request`s are encoded as a type byte followed by the RLP encoding
+    of the request.
+    """
+    if isinstance(req, DepositRequest):
+        return DEPOSIT_REQUEST_TYPE + rlp.encode(req)
+    elif isinstance(req, WithdrawalRequest):
+        return WITHDRAWAL_REQUEST_TYPE + rlp.encode(req)
+    elif isinstance(req, ConsolidationRequest):
+        return CONSOLIDATION_REQUEST_TYPE + rlp.encode(req)
+    else:
+        raise Exception("Unknown request type")
+
+
+def decode_request(data: Bytes) -> Request:
+    """
+    Decode a request.
+    """
+    if data.startswith(DEPOSIT_REQUEST_TYPE):
+        return rlp.decode_to(DepositRequest, data[1:])
```

Comment:
Hm... Looking at this a second time, I'm not sure I like this. To be the _most correct_, we'd need something like:

```suggestion
        return rlp.decode_to(DepositRequest, data[len(DEPOSIT_REQUEST_TYPE):])
```
Hardcoding a length of one would break if `DEPOSIT_REQUEST_TYPE` changed size.

This is a _huge_ nitpick on my part, so if you'd like to ignore it, please do.

---

I guess the alternative would be making `DEPOSIT_REQUEST_TYPE` and friends all integer types, and using something like `return bytes([DEPOSIT_REQUEST_TYPE]) + rlp.encode(req)` above... Not pretty either.

Reply:
This is a valid point imo but it also affects other `encode`/`decode` function pairs such as transaction and receipt encoding. It also goes beyond just the `prague` fork. So I will raise another issue for post-prague re-factoring and take care of it then.